### DISCUSSION TITLE
Red & blue color inverted

### DISCUSSION THREAD

**Discussion by ZhaomingPsych** [sn-lab/MouseGoggles#13] [Q&A] [answered]
Hello everyone, I'm experiencing color issues with fbcp-ili9341. Specifically, I've noticed that red is displaying as yellow, green is displaying as red, and blue is displaying as green. I attempted to resolve this problem by setting "-DDISPLAY_SWAP_BGR=ON" in the config.h file, but it doesn't appear to have made any difference. I would appreciate any assistance or suggestions you may have. Thank you!
![微信图片_20230719234017](https://github.com/sn-lab/mouseVRheadset/assets/96122605/1d961380-4f6b-4f65-956a-6a06539e3153)

**Comment by misaacson01** [comment]
Were these display purchased recently? Because if so, apparently the manufacturer swapped out the driver chip, and this is one of the differences seen in a [recent solved issue](https://github.com/sn-lab/mouseVRheadset/issues/9). It's apparently an easy fix that has been tested, it just needs color inversion turned on in the initialization. 

If you want to make the change yourself, in gc9307.cpp, just after line 66, add the following line:
`SPI_TRANSFER(SPI_CS_BIT, 0x21/*Display Inversion On*/);`
After this change, you'll need to possibly shut down the current driver (step 5 of [operating the system](https://github.com/sn-lab/mouseVRheadset#operating-the-system)), then follow [software installation](https://github.com/sn-lab/mouseVRheadset#software-installation) step 3 to rebuild and reinstall the driver to make the change.

Alternatively if you want to start over with a fresh installation with the change already implemented, you can follow the [software installation](https://github.com/sn-lab/mouseVRheadset#software-installation) instructions, but change step 3 to `cd mouseVRheadset/fbcp-st7789` (which refers to the name of the new chip). But you might not want to do this if you want to preserve any other changes you've made to the Godot files.

Let me know what you see!

**Reply by ZhaomingPsych** [reply, marked as answer]
Thank you so much! I tried method 2, but some issues persisted (shown in the picture). To resolve them, I made a modification in the "config.h" file by changing -DSPI_BUS_CLOCK_DIVISOR=8 and 10 (instead of 6). After implementing these changes, all the problems were successfully resolved!
![微信图片_20230720093112](https://github.com/sn-lab/mouseVRheadset/assets/96122605/2f2a222d-4364-4bb6-bf06-693ac09e5f10)

**Reply by misaacson01** [reply]
I see, thanks for the update! I wonder why this is the case. I know DIVISOR=6 is quite close to the speed limit of these displays, maybe some subtle issues in the PCB assembly can degrade the signal quality just enough to see this issue.

**Reply by ZhaomingPsych** [reply]
Sorry, I'm not sure why this is happening. I consulted the readme file from the author of fbcp-ili9341, and it seems that the driver is updating pixels on the display, but the output appears scrambled or garbled.

**Reply by misaacson01** [reply]
Is this from the same display and PCB that was previously functioning, and that you had to increase DIVISOR=8 to get it working? Because if so, my best guess is still that there is a signal condition problem somewhere. One thing I just remembered is that if you increase the DIVISOR, you probably also need to reduce the target frame rate (to 60, or maybe down to 40 just to see if this helps at all), also in the config.h file. But, I think these are just managing symptoms of poor signals. Maybe you could try reflowing the PCB and/or cleaning up the solder connections, or just replacing it with a new board entirely? This could be a time to start swapping out components to find the culprit.

**Reply by ZhaomingPsych** [reply]
Thank you, I understand your point. I will attempt to solder a new PCB to determine if this issue persists. I'll keep you updated on the progress.
